File: backend/src/main/kotlin/com/mikaservices/platform/modules/projet/entity/Projet.kt
package com.mikaservices.platform.modules.projet.entity

import com.mikaservices.platform.common.entity.BaseEntity
import com.mikaservices.platform.common.enums.ConditionAcces
import com.mikaservices.platform.common.enums.ModeSuiviMensuel
import com.mikaservices.platform.common.enums.SourceFinancement
import com.mikaservices.platform.common.enums.StatutProjet
import com.mikaservices.platform.common.enums.TypeProjet
import com.mikaservices.platform.common.enums.ZoneClimatique
import com.mikaservices.platform.modules.user.entity.User
import jakarta.persistence.*
import java.math.BigDecimal
import java.time.LocalDate

@Entity
@Table(name = "projets", indexes = [
    Index(name = "idx_projet_numero_marche", columnList = "numero_marche"),
    Index(name = "idx_projet_code", columnList = "code_projet"),
    Index(name = "idx_projet_statut", columnList = "statut"),
    Index(name = "idx_projet_client", columnList = "client_id")
])
class Projet(
    @Column(name = "code_projet", nullable = false, unique = true, length = 50)
    var codeProjet: String,

    @Column(name = "numero_marche", length = 100)
    var numeroMarche: String? = null,

    @Column(name = "nom", nullable = false, length = 300)
    var nom: String,

    @Column(name = "description", columnDefinition = "TEXT")
    var description: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, length = 30)
    var type: TypeProjet,

    /** Types multiples (ex. voirie + ouvrage d'art). Le champ type ci‑dessus garde le premier pour index/filtre. */
    @ElementCollection(fetch = FetchType.LAZY)
    @CollectionTable(name = "projet_types", joinColumns = [JoinColumn(name = "projet_id")])
    @Enumerated(EnumType.STRING)
    @Column(name = "type_value", length = 30)
    @org.hibernate.annotations.BatchSize(size = 50)
    var types: MutableSet<TypeProjet> = mutableSetOf(),

    @Column(name = "type_personnalise", length = 150)
    var typePersonnalise: String? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "statut", nullable = false, length = 30)
    var statut: StatutProjet = StatutProjet.EN_ATTENTE,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "client_id")
    var client: Client? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "source_financement", length = 30)
    var sourceFinancement: SourceFinancement? = null,

    @Column(name = "imputation_budgetaire", length = 100)
    var imputationBudgetaire: String? = null,

    @Column(name = "province", length = 100)
    var province: String? = null,

    @Column(name = "ville", length = 100)
    var ville: String? = null,

    @Column(name = "quartier", length = 100)
    var quartier: String? = null,

    @Column(name = "adresse", length = 500)
    var adresse: String? = null,

    @Column(name = "latitude")
    var latitude: Double? = null,

    @Column(name = "longitude")
    var longitude: Double? = null,

    @Column(name = "superficie", precision = 15, scale = 2)
    var superficie: BigDecimal? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "condition_acces", length = 20)
    var conditionAcces: ConditionAcces? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "zone_climatique", length = 20)
    var zoneClimatique: ZoneClimatique? = null,

    @Column(name = "distance_depot_km")
    var distanceDepotKm: Double? = null,

    @Column(name = "nombre_ouvriers_prevu")
    var nombreOuvriersPrevu: Int? = null,

    @Column(name = "horaire_travail", length = 50)
    var horaireTravail: String? = null,

    @Column(name = "montant_ht", precision = 20, scale = 2)
    var montantHT: BigDecimal? = null,

    @Column(name = "montant_ttc", precision = 20, scale = 2)
    var montantTTC: BigDecimal? = null,

    @Column(name = "montant_initial", precision = 20, scale = 2)
    var montantInitial: BigDecimal? = null,

    @Column(name = "montant_revise", precision = 20, scale = 2)
    var montantRevise: BigDecimal? = null,

    @Column(name = "delai_mois")
    var delaiMois: Int? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "mode_suivi_mensuel", length = 10)
    var modeSuiviMensuel: ModeSuiviMensuel? = ModeSuiviMensuel.AUTO,

    @Column(name = "date_debut")
    var dateDebut: LocalDate? = null,

    @Column(name = "date_fin")
    var dateFin: LocalDate? = null,

    @Column(name = "date_debut_reel")
    var dateDebutReel: LocalDate? = null,

    @Column(name = "date_fin_reelle")
    var dateFinReelle: LocalDate? = null,

    @Column(name = "avancement_global", precision = 5, scale = 2)
    var avancementGlobal: BigDecimal = BigDecimal.ZERO,

    @Column(name = "avancement_physique_pct", precision = 5, scale = 2)
    var avancementPhysiquePct: BigDecimal? = null,

    @Column(name = "avancement_financier_pct", precision = 5, scale = 2)
    var avancementFinancierPct: BigDecimal? = null,

    @Column(name = "delai_consomme_pct", precision = 5, scale = 2)
    var delaiConsommePct: BigDecimal? = null,

    @Column(name = "besoins_materiel", columnDefinition = "TEXT")
    var besoinsMateriel: String? = null,

    @Column(name = "besoins_humain", columnDefinition = "TEXT")
    var besoinsHumain: String? = null,

    @Column(name = "observations", columnDefinition = "TEXT")
    var observations: String? = null,

    @Column(name = "propositions_amelioration", columnDefinition = "TEXT")
    var propositionsAmelioration: String? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "responsable_projet_id")
    var responsableProjet: User? = null,

    @Column(name = "partenaire_principal", length = 200)
    var partenairePrincipal: String? = null,

    @Column(name = "actif", nullable = false)
    var actif: Boolean = true,

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "projet_partenaires",
        joinColumns = [JoinColumn(name = "projet_id")],
        inverseJoinColumns = [JoinColumn(name = "partenaire_id")]
    )
    var partenaires: MutableSet<Partenaire> = mutableSetOf(),

    @OneToMany(mappedBy = "projet", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var sousProjets: MutableList<SousProjet> = mutableListOf(),

    @OneToMany(mappedBy = "projet", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var caPrevisionnelsRealises: MutableList<CAPrevisionnelRealise> = mutableListOf(),

    @OneToMany(mappedBy = "projet", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var pointsBloquants: MutableList<PointBloquant> = mutableListOf(),

    @OneToMany(mappedBy = "projet", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var previsions: MutableList<Prevision> = mutableListOf(),

    @OneToMany(mappedBy = "projet", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var revisionsBudget: MutableList<RevisionBudget> = mutableListOf(),

    @OneToMany(mappedBy = "projet", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var avancementEtudes: MutableList<AvancementEtudeProjet> = mutableListOf()
) : BaseEntity() {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Projet) return false
        return id != null && id == other.id
    }

    override fun hashCode(): Int = id?.hashCode() ?: nom.hashCode()

    override fun toString(): String = "Projet(id=$id, nom='$nom')"
}

File: frontend_web/mika-services-frontend/src/types/projet.ts
// ============================================
// Types Projet - MIKA SERVICES
// ============================================

// Enums
export type TypeProjet = 'VOIRIE' | 'ROUTE' | 'CHAUSSEE' | 'PONT' | 'OUVRAGE_ART' | 'BATIMENT' | 'ASSAINISSEMENT' | 'TERRASSEMENT' | 'MIXTE' | 'GENIE_CIVIL' | 'REHABILITATION' | 'AMENAGEMENT' | 'AUTRE'
export type StatutProjet = 'INITIALISATION' | 'EN_COURS_EXECUTION' | 'SUSPENSION' | 'RECEPTION_PROVISOIRE' | 'RECEPTION_DEFINITIVE' | 'EN_AVANCE'
export type TypeClient = 'ETAT_GABON' | 'MINISTERE' | 'COLLECTIVITE' | 'ENTREPRISE_PUBLIQUE' | 'ENTREPRISE_PRIVEE' | 'PARTICULIER'
export type SourceFinancement = 'ETAT_GABONAIS' | 'BGFIBANK' | 'BAD' | 'BM' | 'AFD' | 'PARTENARIAT_PUBLIC_PRIVE' | 'FONDS_PROPRES' | 'MIXTE'
export type TypePartenaire = 'SOUS_TRAITANT' | 'CO_TRAITANT' | 'FOURNISSEUR_STRATEGIQUE' | 'BUREAU_CONTROLE' | 'PARTENAIRE_TECHNIQUE'
export type TypeTravaux = 'TERRASSEMENT' | 'VOIRIE' | 'ASSAINISSEMENT' | 'GENIE_CIVIL' | 'BATIMENT' | 'PONT' | 'OUVRAGE_ART' | 'AMENAGEMENT' | 'REHABILITATION'
export type Priorite = 'BASSE' | 'NORMALE' | 'HAUTE' | 'CRITIQUE' | 'URGENTE'
export type StatutPointBloquant = 'OUVERT' | 'EN_COURS' | 'RESOLU' | 'FERME' | 'ESCALADE'
export type TypePrevision = 'HEBDOMADAIRE' | 'MENSUELLE' | 'TRIMESTRIELLE' | 'PRODUCTION' | 'APPROVISIONNEMENT' | 'RESSOURCES_HUMAINES' | 'MATERIEL'
export type PhaseEtude = 'APS' | 'APD' | 'EXE' | 'GEOTECHNIQUE' | 'HYDRAULIQUE' | 'EIES' | 'PAES'
/** État de validation par l'administration (liste déroulante dans l'avancement des études) */
export type EtatValidationEtude = 'NON_DEPOSE' | 'EN_ATTENTE' | 'EN_COURS' | 'VALIDE' | 'REFUSE'
export type ModeSuiviMensuel = 'AUTO' | 'MANUEL'

export type MotifArretChantier =
  | 'FORTES_PLUIES' | 'INONDATION' | 'SOL_SATURE' | 'GLISSEMENT_TERRAIN' | 'CRUE_RIVIERE' | 'CHALEUR_EXTREME' | 'FOUDRE_ORAGE' | 'VEGETATION_ENVAHISSANTE'
  | 'RETARD_PERMIS_CONSTRUIRE' | 'ABSENCE_AUTORISATION_SOL' | 'NON_CONFORMITE_URBANISME' | 'ARRETE_PREFECTORAL' | 'RETARD_ETUDE_IMPACT' | 'PROBLEME_MARCHE_PUBLIC' | 'CHANGEMENT_GOUVERNEMENT' | 'AUDIT_COUR_COMPTES'
  | 'RETARD_PAIEMENT_MAITRE_OUVRAGE' | 'INSUFFISANCE_TRESORERIE' | 'GEL_BUDGETAIRE' | 'NON_OBTENTION_CAUTION' | 'LITIGE_DECOMPTES' | 'DEPASSEMENT_BUDGET' | 'DEFAILLANCE_SOUS_TRAITANT'
  | 'RUPTURE_STOCK_CIMENT' | 'RETARD_LIVRAISON_PORT' | 'BLOCAGE_DOUANE' | 'INDISPONIBILITE_GRANULATS' | 'PANNE_ENGIN_SANS_PIECES' | 'ROUTES_IMPRATICABLES' | 'PENURIE_CARBURANT' | 'COUPURE_ELECTRICITE' | 'PENURIE_EAU'
  | 'ERREUR_ETUDES_TECHNIQUES' | 'MODIFICATION_PROJET_AVENANT' | 'MALFACON_DEMOLITION' | 'DECOUVERTE_RESEAUX_ENTERRES' | 'SOL_FONDATION_NON_CONFORME' | 'DEFAILLANCE_EQUIPEMENT'
  | 'CONFLIT_FONCIER' | 'OPPOSITION_RIVERAINS' | 'GREVE_OUVRIERS' | 'PENURIE_MAIN_OEUVRE' | 'ACCIDENT_GRAVE' | 'EPIDEMIE_CRISE_SANITAIRE' | 'PROBLEME_VISA_MAIN_OEUVRE'
  | 'INSECURITE_VOL_MATERIAUX' | 'DECOUVERTE_MUNITIONS' | 'ORDRE_AUTORITE_SECURITE'
  | 'ESPECE_PROTEGEE_ZONE_CLASSEE' | 'POLLUTION_ACCIDENTELLE' | 'NON_RESPECT_PRESCRIPTIONS_ENV'
  | 'RESILIATION_MAITRE_OEUVRE' | 'LIQUIDATION_JUDICIAIRE' | 'LITIGE_PARTENAIRES_GROUPEMENT' | 'EXPIRATION_ORDRE_SERVICE'
  | 'AUTRE'

/** Libellés des types de projet (affichage liste, détail, export) */
export const TYPE_PROJET_LABELS: Record<string, string> = {
  VOIRIE: 'Voirie',
  ROUTE: 'Route',
  CHAUSSEE: 'Chaussée',
  PONT: 'Pont',
  OUVRAGE_ART: "Ouvrage d'art",
  BATIMENT: 'Bâtiment',
  ASSAINISSEMENT: 'Assainissement',
  TERRASSEMENT: 'Terrassement',
  MIXTE: 'Mixte',
  GENIE_CIVIL: 'Génie Civil',
  REHABILITATION: 'Réhabilitation',
  AMENAGEMENT: 'Aménagement',
  AUTRE: 'Autre',
}

/** Retourne le libellé à afficher pour un type (ou le type personnalisé si AUTRE) */
function getSingleTypeLabel(type: TypeProjet, typePersonnalise?: string | null): string {
  if (type === 'AUTRE' && typePersonnalise?.trim()) return typePersonnalise.trim()
  return TYPE_PROJET_LABELS[type] ?? type.replace(/_/g, ' ')
}

/** Retourne le libellé à afficher pour le(s) type(s) de projet (sélection multiple). Accepte types (tableau) ou type (simple) pour compatibilité API. */
export function getTypeProjetDisplay(typesOrType: TypeProjet[] | TypeProjet, typePersonnalise?: string | null): string {
  const types = Array.isArray(typesOrType) ? typesOrType : typesOrType ? [typesOrType] : []
  if (types.length === 0) return '—'
  const labels = types.map((t) => getSingleTypeLabel(t, typePersonnalise))
  return labels.join(', ')
}

/** Normalise le champ type(s) d’un projet (API peut renvoyer type ou types). */
export function getProjetTypes(projet: { types?: TypeProjet[]; type?: TypeProjet }): TypeProjet[] {
  if (projet.types && projet.types.length > 0) return projet.types
  if (projet.type) return [projet.type]
  return []
}

// Interfaces
export interface AvancementEtudeProjet {
  id: number
  projetId: number
  phase: PhaseEtude
  avancementPct?: number
  dateDepot?: string
  etatValidation?: string
}

export interface ProjetUserSummary {
  id: number
  nom: string
  prenom: string
  email: string
}

export interface Client {
  id: number
  code: string
  nom: string
  type: TypeClient
  ministere?: string
  telephone?: string
  email?: string
  adresse?: string
  contactPrincipal?: string
  telephoneContact?: string
  actif: boolean
  createdAt?: string
}

export interface Partenaire {
  id: number
  code: string
  nom: string
  type: TypePartenaire
  pays?: string
  telephone?: string
  email?: string
  adresse?: string
  contactPrincipal?: string
  actif: boolean
  createdAt?: string
}

export interface Projet {
  id: number
  codeProjet?: string
  numeroMarche?: string
  nom: string
  description?: string
  /** Types de projet (sélection multiple). Conservé pour compatibilité API : si l’API ne renvoie que type, le front utilise types = [type]. */
  types?: TypeProjet[]
  type?: TypeProjet
  typePersonnalise?: string
  statut: StatutProjet
  client?: Client
  sourceFinancement?: SourceFinancement
  imputationBudgetaire?: string
  province?: string
  ville?: string
  quartier?: string
  latitude?: number
  longitude?: number
  montantHT?: number
  montantTTC?: number
  montantInitial?: number
  montantRevise?: number
  delaiMois?: number
  modeSuiviMensuel?: ModeSuiviMensuel
  dateDebut?: string
  dateFin?: string
  dateDebutReel?: string
  dateFinReelle?: string
  avancementGlobal: number
  avancementPhysiquePct?: number
  avancementFinancierPct?: number
  delaiConsommePct?: number
  besoinsMateriel?: string
  besoinsHumain?: string
  observations?: string
  propositionsAmelioration?: string
  responsableProjet?: ProjetUserSummary
  /** Présent si l’API l’expose (repli si responsableProjet absent). */
  responsableProjetId?: number
  partenairePrincipal?: string
  actif: boolean
  chantierActif: boolean
  motifArretChantier?: MotifArretChantier
  detailArretChantier?: string
  dateArretChantier?: string
  nombrePointsBloquantsOuverts: number
  avancementEtudes?: AvancementEtudeProjet[]
  createdAt?: string
  updatedAt?: string
}

export interface ProjetSummary {
  id: number
  codeProjet?: string
  nom: string
  types?: TypeProjet[]
  type?: TypeProjet
  typePersonnalise?: string
  statut: StatutProjet
  clientNom?: string
  province?: string
  ville?: string
  latitude?: number
  longitude?: number
  montantHT?: number
  avancementGlobal: number
  dateDebut?: string
  dateFin?: string
  responsableNom?: string
  /** Aligné API : pour masquer édition / désactivation si non autorisé */
  responsableProjetId?: number
  /** Nombre d'engins affectés au projet */
  nombreEnginsAffectes?: number
  chantierActif?: boolean
  motifArretChantier?: MotifArretChantier
  dateArretChantier?: string
}

export interface PointBloquant {
  id: number
  projetId: number
  projetNom: string
  titre: string
  description?: string
  priorite: Priorite
  statut: StatutPointBloquant
  detectePar?: ProjetUserSummary
  assigneA?: ProjetUserSummary
  dateDetection: string
  dateResolution?: string
  actionCorrective?: string
  createdAt?: string
  updatedAt?: string
}

export interface CAPrevisionnelRealise {
  id: number
  projetId: number
  mois: number
  annee: number
  caPrevisionnel: number
  caRealise: number
  ecart: number
  avancementCumule: number
}

export interface Prevision {
  id: number
  projetId: number
  projetNom: string
  semaine?: number
  annee: number
  description?: string
  type: TypePrevision
  dateDebut?: string
  dateFin?: string
  avancementPct?: number | null
  createdAt?: string
}

/** Résumé du point projet dans un PV hebdo (pour l’historique). */
export interface PvResumeResponse {
  reunionId: number
  dateReunion: string
  resumeTravauxPrevisions?: string
  pointsBloquantsResume?: string
  besoinsMateriel?: string
  besoinsHumain?: string
  propositionsAmelioration?: string
  avancementPhysiquePct?: number
  avancementFinancierPct?: number
  delaiConsommePct?: number
}

/** Une période (semaine) dans l’historique du projet. */
export interface PeriodeHistoriqueResponse {
  semaine: number
  annee: number
  dateReunion?: string
  previsions: Prevision[]
  pointsBloquants: PointBloquant[]
  pvResume?: PvResumeResponse | null
}

/** Réponse agrégée de l’historique d’un projet. */
export interface ProjetHistoriqueResponse {
  projetId: number
  projetNom: string
  periodes: PeriodeHistoriqueResponse[]
}

export interface RevisionBudget {
  id: number
  projetId: number
  ancienMontant: number
  nouveauMontant: number
  motif?: string
  dateRevision: string
  validePar?: ProjetUserSummary
  createdAt?: string
}

// Request types
export interface ProjetCreateRequest {
  numeroMarche?: string
  nom: string
  description?: string
  /** Types de projet (sélection multiple). Au moins un type requis. */
  types: TypeProjet[]
  typePersonnalise?: string
  statut?: StatutProjet
  clientId?: number
  sourceFinancement?: SourceFinancement
  imputationBudgetaire?: string
  province?: string
  ville?: string
  quartier?: string
  latitude?: number
  longitude?: number
  montantHT?: number
  montantTTC?: number
  montantInitial?: number
  montantRevise?: number
  delaiMois?: number
  modeSuiviMensuel?: ModeSuiviMensuel
  dateDebut?: string
  dateFin?: string
  dateDebutReel?: string
  dateFinReelle?: string
  avancementGlobal?: number
  avancementPhysiquePct?: number
  avancementFinancierPct?: number
  delaiConsommePct?: number
  besoinsMateriel?: string
  besoinsHumain?: string
  observations?: string
  responsableProjetId?: number
  partenairePrincipal?: string
  propositionsAmelioration?: string
  partenaireIds?: number[]
}

export interface ProjetUpdateRequest {
  numeroMarche?: string
  nom?: string
  description?: string
  types?: TypeProjet[]
  typePersonnalise?: string
  statut?: StatutProjet
  clientId?: number
  sourceFinancement?: SourceFinancement
  imputationBudgetaire?: string
  province?: string
  ville?: string
  quartier?: string
  latitude?: number
  longitude?: number
  montantHT?: number
  montantTTC?: number
  montantInitial?: number
  montantRevise?: number
  delaiMois?: number
  modeSuiviMensuel?: ModeSuiviMensuel
  dateDebut?: string
  dateFin?: string
  dateDebutReel?: string
  dateFinReelle?: string
  avancementGlobal?: number
  avancementPhysiquePct?: number
  avancementFinancierPct?: number
  delaiConsommePct?: number
  besoinsMateriel?: string
  besoinsHumain?: string
  observations?: string
  propositionsAmelioration?: string
  responsableProjetId?: number
  partenairePrincipal?: string
  partenaireIds?: number[]
  chantierActif?: boolean
  motifArretChantier?: MotifArretChantier
  detailArretChantier?: string
  dateArretChantier?: string
}

export interface ClientCreateRequest {
  code: string
  nom: string
  type: TypeClient
  ministere?: string
  telephone?: string
  email?: string
  adresse?: string
  contactPrincipal?: string
  telephoneContact?: string
}

export interface ClientUpdateRequest {
  nom?: string
  type?: TypeClient
  ministere?: string
  telephone?: string
  email?: string
  adresse?: string
  contactPrincipal?: string
  telephoneContact?: string
}

export interface PointBloquantCreateRequest {
  projetId: number
  titre: string
  description?: string
  priorite?: Priorite
  detecteParId?: number
  assigneAId?: number
  dateDetection?: string
}

export interface PointBloquantUpdateRequest {
  titre?: string
  description?: string
  priorite?: Priorite
  statut?: StatutPointBloquant
  assigneAId?: number
  dateResolution?: string
  actionCorrective?: string
}

// Page response
export interface PageResponse<T> {
  content: T[]
  totalElements: number
  totalPages: number
  size: number
  number: number
  first: boolean
  last: boolean
}
